nucypher/policy/policies.py

```diff
@@ -576,58 +574,69 @@ def generate_policy_parameters(n: int,
-            else:
-                ether_address = unknown_addresses.popleft()
+        to_check = reservoir.draw(quantity_remaining)
 
-            try:
-                # Check if this is a known node.
-                selected_ursula = self.alice.known_nodes[ether_address]
+        # Sample stakers in a loop and feed them to the learner to check
+        # until we have enough in selected_ursulas`.
 
-            except KeyError:
-                # Unknown Node
-                self.alice.learn_about_specific_nodes({ether_address})  # enter address in learning loop
-                unknown_addresses.append(ether_address)
-                continue
+        start_time = maya.now()
+        new_to_check = to_check
+
+        while True:
+
+            # Check if the sampled addresses are already known.
+            # If we're lucky, we won't have to wait for the learner iteration to finish.
+            known = list(filter(lambda x: x in self.alice.known_nodes, to_check))
+            to_check = list(filter(lambda x: x not in self.alice.known_nodes, to_check))
 
+            known = known[:min(len(known), quantity_remaining)] # we only need so many
+            selected_ursulas.update([self.alice.known_nodes[address] for address in known])
+            quantity_remaining -= len(known)
+
+            if quantity_remaining == 0:
+                break
             else:
-                # Known Node
-                found_ursulas.add(selected_ursula)  # We already knew, or just learned about this ursula
+                new_to_check = reservoir.draw_at_most(quantity_remaining)
+                to_check.extend(new_to_check)
 
-        return found_ursulas
+            # Feed newly sampled stakers to the learner
+            self.alice.learn_about_specific_nodes(new_to_check)
 
-    def sample_essential(self, quantity: int, handpicked_ursulas: Set[Ursula] = None) -> Set[Ursula]:
-        # TODO: Prevent re-sampling of handpicked ursulas.
-        selected_addresses = set()
-        try:
-            sampled_addresses = self.alice.recruit(quantity=quantity,
-                                                   duration=self.duration_periods,
-                                                   additional_ursulas=self.selection_buffer)
-        except StakingEscrowAgent.NotEnoughStakers as e:
-            error = f"Cannot create policy with {quantity} arrangements: {e}"
-            raise self.NotEnoughBlockchainUrsulas(error)
+            # TODO: would be nice to wait for the learner to finish an iteration here,
+            # because if it hasn't, we really have nothing to do.
+            time.sleep(learner_timeout)
+
+            delta = maya.now() - start_time
+            if delta.total_seconds() >= timeout:
+                still_checking = ', '.join(to_check)
+                raise RuntimeError(f"Timed out after {timeout} seconds; "
+                                   f"need {quantity} more, still checking {still_checking}.")
```

Comment:
Should this be `quantity_remaining`?

```suggestion
                                   f"need {quantity_remaining} more, still checking {still_checking}.")
```

Reply:
Thanks, fixed.